device.py

```diff
@@ -20,23 +28,67 @@ def get(self, uuid):
             raise
 
     def get_all(self):
-        return self.request('device', 'GET')
+        return self.request('device', 'GET')['d']
 
     def get_all_by_application(self, name):
-        # notworkingyet
         params = {
-            'expand': 'application',
-            'eq': name,
-            'filter': "app_name"
+            'filter': 'app_name',
+            'eq': name
         }
-        return self.request('device', 'GET', params=params)
+        app = self.request('application', 'GET', params=params)['d']
+        print app
+        if app:
+            params = {
+                'filter': 'application ',
+                'eq': app[0]['id']
+            }
+            return self.request('device', 'GET', params=params)['d']
 
     def get_by_name(self, name):
         params = {
             'filter': 'name',
             'eq': name
         }
-        return self.request('device', 'GET', params=params)
+        return self.request('device', 'GET', params=params)['d']
+
+    def get_name(self, uuid):
+        return self.get(uuid)['name']
+
+    def get_application_name(self, uuid):
+        app_id = self.get(uuid)['application']['__id']
+        params = {
+            'filter': 'id',
+            'eq': app_id
+        }
+        return self.request('application', 'GET', params=params)['app_name']
+        
+    def has(self, uuid):
+        params = {
+            'filter': 'uuid',
+            'eq': uuid
+        }
+        if len(self.request('device', 'GET', params=params)) > 0:
+            return True
+        return False 
+
+    def is_online(self, uuid):
+        return self.get(uuid)['is_online']
+
+    def get_local_ip_address(self, uuid):
+        if self.is_online(uuid):
+            device = self.get(uuid)
+            ips = device['ip_address'].split()
+            ips.remove(device['vpn_address'])
+            return ips
+        else:
+            print(Message.DEVICE_OFFLINE.format(uuid=uuid))
```

Comment:
This method is called ``get_local_ip_address()`` and returns either a list of somethings or nothing (but prints a message to the console). Imagine how many different cases you'd have to handle if you were a user of the SDK and all you wanted was to get your device's IP.